File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/internal/DriverFleetBindingsMapper.kt
package com.mgl.fleet.sdk.internal

import com.mgl.fleet.sdk.demo.DemoAuthMode
import com.mgl.fleet.sdk.demo.DemoBinding
import com.mgl.fleet.sdk.demo.DemoBindingState
import kotlin.math.roundToLong
import java.util.Locale as JLocale

internal fun normVrnPublic(v: String): String =
    v.trim().replace("\\s+".toRegex(), " ").uppercase(JLocale.ENGLISH)

private fun parseShiftDays(csv: String?): List<String>? {
    if (csv.isNullOrBlank()) return null
    return csv.split(',').map { d ->
        val t = d.trim()
        if (t.length < 2) t else t.take(1) + t.substring(1, 3.coerceAtMost(t.length)).lowercase(JLocale.ENGLISH)
    }
}

private fun assignmentTypeToAuthMode(t: String): DemoAuthMode =
    when (t.uppercase(JLocale.ENGLISH)) {
        "WHOLE_TIME" -> DemoAuthMode.VEHICLE_LINKED
        "SHIFT" -> DemoAuthMode.SHIFT_BASED
        else -> DemoAuthMode.TRIP_LINKED
    }

/** Mirrors `mapAssignmentsToUiBindings` from `driver-api.ts` into [DemoBinding]. */
internal fun mapAssignmentsToDemoBindings(
    home: DriverHomeJson?,
    rows: List<DriverAssignmentJson>,
): List<DemoBinding> {
    val homeVrn = home?.vehicleRegNo?.let { normVrnPublic(it) }.orEmpty()
    val hasHomeVrn = homeVrn.isNotEmpty()
    return rows.map { a ->
        val authMode = assignmentTypeToAuthMode(a.assignmentType)
        val vrnKey = normVrnPublic(a.vehicleRegNo)
        val matchesHome = home?.hasActiveVehicle == true && hasHomeVrn && vrnKey == homeVrn
        val activeEligible = a.status == "ACTIVE" && !a.requiresPairing
        val homeEligible =
            home?.isCurrentlyEligible == true ||
                home?.currentlyEligible == true ||
                (
                    home?.isCurrentlyEligible == null &&
                        home?.currentlyEligible == null &&
                        activeEligible
                    )
        val eligible = if (matchesHome) homeEligible else activeEligible

        val scanPay =
            when {
                a.status == "PENDING_ACCEPTANCE" && a.requiresPairing -> "locked_unpaired"
                !eligible -> "out_window"
                a.assignmentType.uppercase(JLocale.ENGLISH) == "WHOLE_TIME" -> "always_available"
                a.assignmentType.uppercase(JLocale.ENGLISH) == "SHIFT" -> "in_window"
                else -> "trip_window"
            }

        val balance: Long =
            if (matchesHome && home?.totalBalanceINR != null) {
                home.totalBalanceINR!!.roundToLong()
            } else {
                0L
            }
        val foFromHome =
            if (matchesHome && !home?.foName.isNullOrBlank()) {
                home!!.foName!!.trim()
            } else {
                ""
            }

        val state =
            if (a.status == "PENDING_ACCEPTANCE") {
                DemoBindingState.PENDING_ACCEPTANCE
            } else {
                DemoBindingState.ACTIVE
            }
        val paired = !(a.status == "PENDING_ACCEPTANCE" && a.requiresPairing)

        DemoBinding(
            id = a.vehicleDriverId.toString(),
            vrn = a.vehicleRegNo,
            fo = foFromHome,
            vehicleId = a.vehicleId,
            authMode = authMode,
            state = state,
            paired = paired,
            scanPayStatus = scanPay,
            shiftDays = parseShiftDays(a.shiftDaysOfWeek) ?: emptyList(),
            shiftStart = a.shiftStartTime ?: "",
            shiftEnd = a.shiftEndTime ?: "",
            tripDate = a.tripDate ?: "",
            tripStart = a.tripStartTime ?: "",
            tripEnd = a.tripEndTime ?: "",
            origin = a.tripStartLocation ?: "",
            destination = "",
            balance = balance,
            cardBalance = balance,
            incentiveBalance = 0,
        )
    }
}

File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/DriverFleetBindingsMapper.swift
import Foundation

enum DriverFleetBindingsMapper {
    static func mapAssignmentsToDemoBindings(
        home: DriverHomeParsed?,
        rows: [DriverAssignmentParsed],
    ) -> [DemoBinding] {
        let homeVrn = home?.vehicleRegNo.map { DriverFleetQr.normVrnPublic($0) } ?? ""
        let hasHomeVrn = !homeVrn.isEmpty

        func parseShiftDays(_ csv: String?) -> [String]? {
            guard let csv, !csv.isEmpty else { return nil }
            return csv.split(separator: ",").map { d in
                let t = String(d).trimmingCharacters(in: .whitespaces)
                guard t.count >= 2 else { return t }
                let take = min(3, t.count)
                let idx = t.index(t.startIndex, offsetBy: take)
                let suffix = String(t[t.index(after: t.startIndex) ..< idx]).lowercased(with: Locale(identifier: "en_US"))
                return String(t.prefix(1)) + suffix
            }
        }

        func authMode(for t: String) -> DemoAuthMode {
            switch t.uppercased() {
            case "WHOLE_TIME": return .vehicleLinked
            case "SHIFT": return .shiftBased
            default: return .tripLinked
            }
        }

        return rows.map { a in
            let matchesHome = home?.hasActiveVehicle == true && hasHomeVrn &&
                DriverFleetQr.normVrnPublic(a.vehicleRegNo) == homeVrn
            let activeEligible = a.status == "ACTIVE" && !a.requiresPairing
            let homeEligible =
                home?.isCurrentlyEligible == true
                || home?.currentlyEligible == true
                || (home?.isCurrentlyEligible == nil && home?.currentlyEligible == nil && activeEligible)
            let eligible = matchesHome ? homeEligible : activeEligible

            let scanPay: String = {
                if a.status == "PENDING_ACCEPTANCE" && a.requiresPairing { return "locked_unpaired" }
                if !eligible { return "out_window" }
                switch a.assignmentType.uppercased() {
                case "WHOLE_TIME": return "always_available"
                case "SHIFT": return "in_window"
                default: return "trip_window"
                }
            }()

            let balance: Int =
                matchesHome ? Int((home?.totalBalanceINR ?? 0).rounded()) : 0
            let foFromHome: String = {
                guard matchesHome, let n = home?.foName?.trimmingCharacters(in: .whitespaces), !n.isEmpty else { return "" }
                return n
            }()

            let state: DemoBindingState = a.status == "PENDING_ACCEPTANCE" ? .pendingAcceptance : .active
            let paired = !(a.status == "PENDING_ACCEPTANCE" && a.requiresPairing)

            return DemoBinding(
                id: String(a.vehicleDriverId),
                vrn: a.vehicleRegNo,
                fo: foFromHome,
                vehicleId: a.vehicleId,
                authMode: authMode(for: a.assignmentType),
                state: state,
                paired: paired,
                scanPayStatus: scanPay,
                balance: balance,
                cardBalance: balance,
                incentiveBalance: 0,
                spendLimit: 2000,
                shiftDays: parseShiftDays(a.shiftDaysOfWeek) ?? [],
                shiftStart: a.shiftStartTime ?? "",
                shiftEnd: a.shiftEndTime ?? "",
                tripStart: a.tripStartTime ?? "",
                tripEnd: a.tripEndTime ?? "",
                tripDate: a.tripDate ?? "",
                origin: a.tripStartLocation ?? "",
                destination: "",
                assignedBy: nil,
                validPairingCode: nil,
                repairReason: nil,
            )
        }
    }
}
